text/0000-freezable-mutex-guards.md

```diff
@@ -0,0 +1,191 @@
+to release the lock. This often leads to call chains where every method requires the owned Guard and the Mutex reference
+as parameter and the owned Guard as return type. This produces unreadable code and leads to a lot of useless parameter
+passing. The second approach if passing the Mutex is not wanted is to lock and unlock it in every scope needed. With
+this style the code is much more readable, but it takes a lot of unnecessary locks and unlocks of the Mutex. This
+approach is less performant.
+
+The concept is to add a method with three parameters for the guard. First, obviously the mutable reference to the guard.
+The second parameter is a closure containing code that should be run while the Mutex is unlocked. This by itself would
+already work, but it might cause problems if another thread panics while holding the data. Therefore, the third
+parameter is a "heal" closure. It takes a `PoisonError` with the guard as parameter to allow the thread to fix the inner
+data.
+
+This would allow always just using the mutable reference instead of the owned object. Which leads to better readablity
+and maintainability while still being safe.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Unlocks the mutex temporary and lock it after the calling `func` again. If the Mutex is poisoned upon trying to lock it 
+again, the `heal` function will be called to fix the inner data. After this, the poison will be cleared and the mutex 
+guard is valid again.
+
+### Example
+
+```rust
+use std::thread;
+use std::sync::Mutex;
+use std::sync::Condvar;
+
+pub fn request_data(guard: &mut MutexGuard<usize>, condvar: &Condvar) {
+    guard.freeze_guard(
+        || { &condvar.notify_one(); },
+        |poison| { panic!() }
+    );
+}
+
+pub fn main() {
+    let mutex = Mutex::new(0usize);
+    let condvar = Condvar::new();
+
+    thread::scope(|s| {
+        let outer_guard = mutex.lock().unwrap();
+
+        s.spawn(|| {
+            let inner_guard = mutex.lock().unwrap();
+            
+            let inner_guard_ref = &mut inner_guard;
+            assert_eq!(**inner_guard_ref, 0);
+            request_data(inner_guard_ref, &condvar);
+            assert_eq!(**inner_guard_ref, 123);
+
+        });
+
+        let outer_guard = &condvar.wait(outer_guard).unwrap();
+        **outer_guard = 123;
+        drop(outer_guard)
+    })
+}
+```
+
+### Panics:
+
+If the thread panics while running the `func` closure the mutex will remain clear because it is not locked by the
+current thread (excluding the case that the thread is acquiring the lock in the `func`). If the thread panics in the
+`heal` function the thread mutex remains poisoned.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+impl<'a, T: ?Sized> MutexGuard<'a, T> {
+    pub fn freeze_guard<F, H>(&mut self, func: F, heal: H) -> () 
```

Comment:
```suggestion
    pub fn freeze_guard<F, H>(orig: &mut Self, func: F, heal: H) -> () 
```

Similar to `MutexGuard::{map, filter_map}` this has to be used as `MutexGuard::freeze_guard(guard, ...)` to avoid interfering with Deref.